src/Servers/Kestrel/Core/src/Internal/Http/HttpParser.cs

```diff
@@ -306,44 +325,77 @@ private int ParseMultiSpanHeader(TRequestHandler handler, ref SequenceReader<byt
 
         SequencePosition lineEnd;
         scoped ReadOnlySpan<byte> headerSpan;
+        ReadOnlySequence<byte> header;
         if (currentSlice.Slice(reader.Position, lineEndPosition.Value).Length == currentSlice.Length - 1)
         {
-            // No enough data, so CRLF can't currently be there.
-            // However, we need to check the found char is CR and not LF
+            // If we're not allowing LF by itself as a terminator, we need two characters
+            // for the line terminator. Since we don't have two, we know there can't be 
+            // a CRLF here. However, we need to also check that found char is CR and not LF.
 
             // Advance 1 to include CR/LF in lineEnd
             lineEnd = currentSlice.GetPosition(1, lineEndPosition.Value);
-            var header = currentSlice.Slice(reader.Position, lineEnd);
-            headerSpan = header.IsSingleSegment ? header.FirstSpan : header.ToArray();
-            if (headerSpan[^1] != ByteCR)
+            header = currentSlice.Slice(reader.Position, lineEnd);
+            headerSpan = header.ToSpan();
+            if (headerSpan[^1] == ByteCR)
             {
-                RejectRequestHeader(headerSpan);
+                // CR without LF, can't read the header
+                return -1;
             }
-            return -1;
+            else
+            {
+                if (_disableHttp1LineFeedTerminators)
+                {
+                    // LF and disabled LF
+                    RejectRequestHeader(headerSpan);
+                }
+            }
+        }
+
+        // Offset 1 to include the first line end char.
+        var firstLineEndCharPos = currentSlice.GetPosition(1, lineEndPosition.Value);
+        header = currentSlice.Slice(reader.Position, firstLineEndCharPos);
+
+        headerSpan = header.ToSpan();
+
+        if (headerSpan[^1] == ByteCR)
+        {
+            header = currentSlice.Slice(reader.Position, currentSlice.GetPosition(2, lineEndPosition.Value));
+        }
+        else if (_disableHttp1LineFeedTerminators)
+        {
+            // The terminator is an LF and we don't allow it.
+            RejectRequestHeader(headerSpan);
+        }
+        else
+        {
+            header = currentSlice.Slice(reader.Position, currentSlice.GetPosition(1, lineEndPosition.Value));
         }
 
-        // Advance 2 to include CR{LF?} in lineEnd
-        lineEnd = currentSlice.GetPosition(2, lineEndPosition.Value);
-        headerSpan = currentSlice.Slice(reader.Position, lineEnd).ToSpan();
+        headerSpan = header.ToSpan();
```

Comment:
And this is a second (or sometimes even third) array allocation in this method. I understand that the original `headerSpan` doesn't have the last byte that we need to check, so we should start with `currentSlice.Slice(reader.Position, currentSlice.GetPosition(2, lineEndPosition.Value)).ToSpan()` if possible and pare down from there.